core/src/main/java/org/apache/iceberg/spark/actions/compaction/BinningCompactionStrategy.java

```diff
@@ -0,0 +1,91 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.actions.compaction;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.actions.compaction.DataCompactionStrategy;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.util.BinPacking;
+import org.apache.iceberg.util.PropertyUtil;
+
+import static org.apache.iceberg.util.StreamUtil.streamOf;
+
+/**
+ * A Compaction strategy for rewriting files based on their sizes, the goal is to end up with files
+ * whose size is close to the Target Size as specified in this strategy's options.
+ */
+public abstract class BinningCompactionStrategy implements DataCompactionStrategy {
+
+
+  public static final String TARGET_SIZE_OPTION = "target_size";
+  public static final String TARGET_THRESHOLD_OPTION = "target_threshold";
+
+  // TODO Maybe these should be global?
+  public static final String MAX_JOB_SIZE_OPTION = "max_job_size";
+
+  protected static final long TARGET_THRESHOLD_DEFAULT = 50 * 1024 * 1024; // 50 MB
+  protected static final long TARGET_SIZE_DEFAULT = 512 * 1024 * 1024; // 512 MB // TODO (Table default)
+  protected static final long MAX_JOB_SIZE_DEFAULT = 100L * 1024 * 1024 * 1024; // 100 GB
+
+  private long targetThreshold = TARGET_THRESHOLD_DEFAULT;
+  private long targetSize = TARGET_SIZE_DEFAULT;
+  private long maxJobSize = MAX_JOB_SIZE_DEFAULT;
+
+  @Override
+  public String name() {
+    return "Binning Size Based Compaction Strategy";
+  }
+
+  @Override
+  public Set<String> validOptions() {
+    return ImmutableSet.of(TARGET_SIZE_OPTION, TARGET_THRESHOLD_OPTION, MAX_JOB_SIZE_OPTION);
+  }
+
+  @Override
+  public Iterator<FileScanTask> filesToCompact(Iterator<FileScanTask> dataFiles) {
+    return streamOf(dataFiles)
+        .filter(task -> Math.abs(task.file().fileSizeInBytes() - targetSize) > targetThreshold)
+        .iterator();
+  }
+
+  @Override
+  public Iterator<List<FileScanTask>> groupsFilesIntoJobs(Iterator<FileScanTask> dataFiles) {
+    BinPacking.ListPacker<FileScanTask> packer = new BinPacking.ListPacker<>(maxJobSize, 1, false);
```

Comment:
I think we will need to discuss how to pack files and whether we need to look at the min/max stats on the sort key, for example. Does not have to be done now, though.

Reply:
Yep, I'm assuming some algorithms may want to do that all the time.